### PR TITLE
[Solid]  Integration Schemes for SolidMechanicsApplication and PfemApplication

### PR DESCRIPTION
**Pull request by josep-m-carbonell**
The integration schemes for the Solid and Pfem applications now moved to the SolversApplication. It is an independent application with a specialized implementation of the non-linear strategies, builders and schemes. The files are not new, just moved from the SolidMechanicsApplication to the SolversApplication.
No trace of external library interface.

### CODE REVIEW

**Review by loumalouomega — CHANGES_REQUESTED** [review]
This PR adds tetgen submodule. Can we avoid adding submodules please?

**Comment by josep-m-carbonell** [comment]
thank you Vicente, I have not seen that.

**Review by loumalouomega — DISMISSED** [review]

**Review by loumalouomega — COMMENTED** [review]
Better just comment

File: applications/SolversApplication/SolversApplication.py
Comment (philbucher) [inline applications/SolversApplication/SolversApplication.py:7]: please use the new style for importing applications(as python-modules only), see the template-application 

This will make things easier when Kratos moves to this entirely

**Review by philbucher — CHANGES_REQUESTED** [review]

File: applications/SolversApplication/SolversApplication.py
Comment (josep-m-carbonell) [reply applications/SolversApplication/SolversApplication.py:7]: Hello @philbucher, I didn't know about this new change in importing applicacions. If it is ok for you I will do another PR updating all applications I maintain with an update of this file.

File: applications/SolversApplication/python_scripts/monolithic_solver.py
Comment (philbucher) [inline applications/SolversApplication/python_scripts/monolithic_solver.py]: This is exactly what the @KratosMultiphysics/technical-committee said they don't want

File: scripts/build/appveyor/configure_appveyor_VS2017.bat
Comment (philbucher) [inline scripts/build/appveyor/configure_appveyor_VS2017.bat:14]: Applications cannot be added to the CI before approval of the @KratosMultiphysics/technical-committee , even if you remove others

The CI is already heavily overloaded and this affects everyone

**Review by philbucher — CHANGES_REQUESTED** [review]
@josep-m-carbonell 
I looked some more into this PR and in my role as a member of @KratosMultiphysics/design-committee I think this does not fit with the global ideas/goals of Kratos in several points:
- An application called `SolversApplication` sounds very general and should be usable from many other applications. At the moment I don't see this possible, it seems to only work for the SolidMechanicsApp. This is also really confusing for new users of Kratos.
- The @KratosMultiphysics/technical-committee said several times already that they want to go for fewer larger and well maintained/consolidated Applications rather than many small apps. Can you please give an explanation of why you want to separate the apps?
- The things you add to this application seem to be a duplication of code from the Core (I see some of them are only moved from Solid)!
    - Why do you need e.g. your own BuilderAndSolvers? The ones in the core should be preferred to be used, and only in really special cases a custom B&S is needed. If there are some things not working for the things in the Core, then it would be better if you point those out, such that they can be improved and the entire Kratos-community can benefit
    - **As a meanwhile quite large open-source project it makes us look unprofessional if we have the same capabilities implemented several times in many places!**  I really think we should consolidate ourselves rather than let the code deviate more and more, otherwise Kratos will not live for much longer I fear.
    - Maintenance becomes much harder if code is duplicated
    - I myself and many others are trying really hard to refactor and reduce the duplication of code such that the entire Community can benefit from it
    - There are **no tests** added for the new application! It adds many complicated things that really require some proper testing in C and Python

@KratosMultiphysics/technical-committee I think you should also review this PR, I pointed out many things that need a review / approval from you

**Comment by loumalouomega** [comment]
I suscribe the conclusions of this review. Speaking as member of the
Implementation Committee

**Comment by josep-m-carbonell** [comment]
image

File: applications/SolversApplication/custom_solvers/solution_schemes/eigensolver_scheme.hpp
Comment (RiccardoRossi) [inline applications/SolversApplication/custom_solvers/solution_schemes/eigensolver_scheme.hpp]: please look into the discussion #2452 for guidelines on autorship.

If the change here is only in the include, it is a minimal modification, so you should not put yourself as author of this (even if you acknolwedge the source)
Also note that if you found errors in the original implementation it would be fair to contribute back the corrections

File: applications/SolversApplication/custom_solvers/solution_schemes/explicit_central_differences_scheme.hpp
Comment (RiccardoRossi) [inline applications/SolversApplication/custom_solvers/solution_schemes/explicit_central_differences_scheme.hpp]: see policy

File: applications/SolversApplication/python_scripts/monolithic_solver.py
Comment (RiccardoRossi) [reply applications/SolversApplication/python_scripts/monolithic_solver.py]: thx @philbucher for spotting

**Review by RiccardoRossi — COMMENTED** [review]
Aside for the sparse comments in the code,
I think that calling the new application "SolversApp" is misleading

The new application is deliberately incompatible with the rest of core and with many of the core applications. 
Please name it as "SolidSolversApplication" or something of the sort to make it clear that it belongs to your constellation of apps and is not compatible with the rest of the applications

File: applications/SolversApplication/custom_solvers/solution_schemes/eigensolver_scheme.hpp
Comment (josep-m-carbonell) [reply applications/SolversApplication/custom_solvers/solution_schemes/eigensolver_scheme.hpp]: It is the same in the structural mechanics application. Which is a bad copy of the solid application.

File: applications/SolversApplication/custom_solvers/solution_schemes/explicit_central_differences_scheme.hpp
Comment (josep-m-carbonell) [reply applications/SolversApplication/custom_solvers/solution_schemes/explicit_central_differences_scheme.hpp]: Miquel is ok with this.

**Comment by josep-m-carbonell** [comment]
Now you have problems with names Mr. "NOT DONOT_NDEBUG?"

**Comment by RiccardoRossi** [comment]
I prefer not to answer.

File: applications/SolversApplication/custom_solvers/solution_schemes/eigensolver_scheme.hpp
Comment (maceligueta) [reply applications/SolversApplication/custom_solvers/solution_schemes/eigensolver_scheme.hpp]: Can we focus on this PR? If things were incorrectly done in the past (arguable), that is no reason to keep doing them wrong.

Comment (philbucher) [reply applications/SolversApplication/custom_solvers/solution_schemes/eigensolver_scheme.hpp]: It is the one from StructuralMechanics, which was migrated to SolidMechanics in #736

The original files were added here: https://github.com/KratosMultiphysics/Kratos/commit/39193d50de7053516f39f6c1e75da9976fc4ea6b#diff-ae31bd10d286705c1ab8892043f27b7a
(almost one year before it was copied to Solid...)

Comment (josep-m-carbonell) [reply applications/SolversApplication/custom_solvers/solution_schemes/eigensolver_scheme.hpp]: I took this as reference:
mechanical_explicit_strategy.hpp
//  Main authors:    Klaus B Sautter (based on the work of JMCarbonel)
explicit_central_differences_scheme.hpp:
//  Main authors:    Klaus B Sautter (based on the work of MSantasusana)

Comment (philbucher) [reply applications/SolversApplication/custom_solvers/solution_schemes/eigensolver_scheme.hpp]: There it was properly referenced where the files came from

Also these files were heavily modified to account for the rotations (e.g. of shell & beam-elements)

File: applications/SolversApplication/custom_solvers/solution_schemes/explicit_central_differences_scheme.hpp
Comment (josep-m-carbonell) [reply applications/SolversApplication/custom_solvers/solution_schemes/explicit_central_differences_scheme.hpp]: I updated the headers to avoid conflicts and including all names.

File: applications/SolversApplication/custom_solvers/solution_schemes/eigensolver_scheme.hpp
Comment (josep-m-carbonell) [reply applications/SolversApplication/custom_solvers/solution_schemes/eigensolver_scheme.hpp]: The reference it was exactly the same. It sounds like, when I do myself is ok, when others do it is wrong
I have updated the headers to avoid these conflicts. 

P.D. There is a typo in (based on the work of JMCarbonel), my surname is Carbonell.

**Comment by josep-m-carbonell** [comment]
It seems that custom developments and applications not belonging to "committees" are banned. The restrictions imposed affect directly to the development of the work of researchers, penalizing their tasks. It is clear that the Kratos project in this repository has taken the way to end up being a semi-private property of consortioum TUM-CIMNE / Altair.  If the developments of any researcher are not in the line of the interests of this consortium, they are not welcome. It is a pity. I am afraid it will be the start of the end. I hope I am wrong. @KratosMultiphysics/all

**Comment by pooyan-dadvand** [comment]
**PR and issues are technical forums** and should be used as such. In addition, using **offensive language and personal disqualification is unacceptable**. Please follow these rules when commenting in these channels. Repeated violations of these rules will result in messages moderation and blocking the account.

If you have any comment or concerns about the community and its way of working (which is accepted by most of the developers) we are open to listening to you in a meeting which can be open to anyone from the Kratos community.

**Pull request closed without merging**
